crates/iota-core/src/authority/suggested_gas_price_calculator.rs

```diff
@@ -0,0 +1,1255 @@
+            update_data_for_scheduled_certificate(
+                &certificate,
+                execution_start_time,
+                &mut shared_object_congestion_tracker,
+                &mut suggested_gas_price_calculator,
+            );
+        } else {
+            panic!(
+                "Certificate {} must be scheduled",
+                tx_gas_data.global_ordering_index
+            );
+        }
+
+        // Construct a certificate with some shared objects (note mutability),
+        // and try scheduling it.
+        let input_shared_objects = vec![(object_1, false), (object_2, true)];
+        let tx_gas_data = *txs_gas_data.get(1).unwrap();
+        let (certificate, sequencing_result) = build_and_try_sequencing_certificate(
+            &input_shared_objects,
+            tx_gas_data.gas_price,
+            tx_gas_data.gas_budget,
+            max_execution_duration_per_commit,
+            &mut shared_object_congestion_tracker,
+        );
+        // Allocations of mutably accessed shared objects should look as follows:
+        // |----------------------------------------------------
+        // |        object_1         |        object_2         |
+        // |_________________________|_________________________|
+        // |-------------------------|-------------------------|
+        // |                         |                         |
+        // |-------------------------|-------------------------|
+        // |                         | cert. 1 (g=9000, d=10M) |
+        // |-------------------------|-------------------------|
+        // | cert. 0 (g=100K, d=12M) |                         |
+        // |----------------------------------------------------
+        if let SequencingResult::Schedule(execution_start_time) = sequencing_result {
+            update_data_for_scheduled_certificate(
+                &certificate,
+                execution_start_time,
+                &mut shared_object_congestion_tracker,
+                &mut suggested_gas_price_calculator,
+            );
+        } else {
+            panic!(
+                "Certificate {} must be scheduled",
+                tx_gas_data.global_ordering_index
+            );
+        }
+
+        // Construct a certificate with some shared objects (note mutability),
+        // and try scheduling it.
+        let input_shared_objects = vec![(object_1, false), (object_2, true)];
+        let tx_gas_data = *txs_gas_data.get(2).unwrap();
+        let (certificate, sequencing_result) = build_and_try_sequencing_certificate(
+            &input_shared_objects,
+            tx_gas_data.gas_price,
+            tx_gas_data.gas_budget,
+            max_execution_duration_per_commit,
+            &mut shared_object_congestion_tracker,
+        );
+        // Allocations of mutably accessed shared objects should look as follows:
+        // |----------------------------------------------------
```

Comment:
These axes are reversed compared with the illustrations for the tests in the shared object congestion tracker. They should ideally be the same for consistency between the tests, but even if you want to keep them different, you should label the y axis here somehow to show that start times descend as we go down the figure.